server/deviceflowhandlers_test.go

```diff
@@ -0,0 +1,672 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"path"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/dexidp/dex/storage"
+)
+
+func TestDeviceVerificationURI(t *testing.T) {
+	t0 := time.Now()
+
+	now := func() time.Time { return t0 }
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// Setup a dex server.
+	httpServer, s := newTestServer(ctx, t, func(c *Config) {
+		c.Issuer = c.Issuer + "/non-root-path"
+		c.Now = now
+	})
+	defer httpServer.Close()
+
+	u, err := url.Parse(s.issuerURL.String())
+	if err != nil {
+		t.Errorf("Could not parse issuer URL %v", err)
+	}
+	u.Path = path.Join(u.Path, "/device/auth/verify_code")
+
+	uri := s.getDeviceVerificationURI()
+	if uri != u.Path {
+		t.Errorf("Invalid verification URI.  Expected %v got %v", u.Path, uri)
+	}
+}
+
+func TestHandleDeviceCode(t *testing.T) {
+	t0 := time.Now()
+
+	now := func() time.Time { return t0 }
+
+	tests := []struct {
+		testName               string
+		clientID               string
+		scopes                 []string
+		expectedResponseCode   int
+		expectedServerResponse string
+	}{
+		{
+			testName:             "New Valid Code",
+			clientID:             "test",
+			scopes:               []string{"openid", "profile", "email"},
+			expectedResponseCode: http.StatusOK,
+		},
+	}
```

Comment:
Could we add some more cases here? Eg, what happens with an invalid code? `expectedServerResponse` is currently unused?

Reply:
The name is a bit misleading.  All this test does is hit the /device/code endpoint, and validate we can get a device code response (device code, user code, verfication uri, expire time, and request limit) in the proper format.  

Ive updated the tests to include a test case for using an invalid request type (GET vs POST) for code coverage, completeness though